server/src/test/java/org/elasticsearch/transport/RelaxedSingleResultDeduplicatorTests.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.RelaxedSingleResultDeduplicator;
+import org.elasticsearch.action.SingleResultDeduplicator;
+import org.elasticsearch.action.support.ActionTestUtils;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+public class RelaxedSingleResultDeduplicatorTests extends SingleResultDeduplicatorTests {
+
+    @Override
+    protected <T> SingleResultDeduplicator<T> makeSingleResultDeduplicator(
+        ThreadContext threadContext,
+        Consumer<ActionListener<T>> executeAction
+    ) {
+        return new RelaxedSingleResultDeduplicator<T>(threadContext, executeAction);
+    }
+
+    public void testDeduplicate() {
+        final int numThreads = between(2, 10);
+        final var actionExecutionCount = new AtomicInteger(0);
+        final var responses = new Object[numThreads];
+        final var resultObjectRef = new AtomicReference<Object>();
+        final var countDownLatch = new CountDownLatch(numThreads - 1);
+
+        final Consumer<ActionListener<Object>> computation = l -> {
+            final var count = actionExecutionCount.incrementAndGet();
+            // The first thread will block until all the other callers have added a waiting listener.
+            safeAwait(countDownLatch);
```

Comment:
This is ok here but we should also have tests that don't orchestrate the order of events quite so carefully so that we can be more confident there's no races that might lead to lost listeners or duplicate concurrent executions. Admittedly the testing of `SingleResultDeduplicator` today is rather weak in this area, I'd be happy to see some more stringent testing of its invariants under concurrent use too.